File: W15_CIS*3190 Legacy Programming/Assignment4/iterative_ackermann.f90
! filename:   iterative_ackermann.f90
! author:     Kyle Hersey
! student#:   0851973
! course:     CIS*3190
! assignment: 4

! main function
Program iterative_ackermann
implicit none

integer, dimension(1000000) :: st
integer :: top, m, n

top = 0

write(*,*) 'Enter M: '
read(*,*) m
write(*,*) 'Enter N: '
read(*,*) n

call ackermann(m, n, st, top)

write(*,*) 'Result is', n

end Program iterative_ackermann


! ackermann function
subroutine ackermann(m, n, st, top)
implicit none

integer, dimension(1000000), intent(in out) :: st
integer, intent(in out) :: top, m, n

call push(m, st, top)
do while (top /= 0)
    call pop(m, st, top)
    if (m == 0) then
        n = n + 1
    else if (n == 0) then
        n = 1
        call push(m - 1, st, top)
    else
        n = n - 1
        call push(m - 1, st, top)
        call push(m, st, top)
    end if
end do

end subroutine ackermann


! stack push operation
subroutine push(item, st, top) 
implicit none

integer, dimension(1000000), intent(in out) :: st
integer, intent(in out) :: top
integer, intent(in) :: item

top = top + 1
if (top == 1000000) then
    write(*,*) 'Error: Stack Overflow'
    call exit(0)
end if

st(top) = item

end subroutine push

! stack pop operation
subroutine pop(item, st, top)
implicit none

integer, dimension(1000000), intent(in out) :: st
integer, intent(in out) :: top
integer, intent(out) :: item


if (top == 0) then
    write(*,*) 'Error: Stack Underflow'
    call exit(0)
end if

item = st(top)
top = top - 1

end subroutine pop

File: W15_CIS*3190 Legacy Programming/Assignment4/iterative_ackermann.c
/* filename:   iterative_ackermann.c
 * author:     Michael Wirth
 *             (From Courselink)
 * course:     CIS*3190
 * assignment: 4
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <sys/time.h>

#define STACKSIZE 1000000

/* Iterative Ackermann from Courselink */

struct dataT {
    int m;
};

struct stack {
    int top;
    struct dataT items[STACKSIZE];
} st;

void createstack() {
    st.top = -1;
}

int isFULL() {
    if (st.top >= STACKSIZE-1) {
        return 1;
    } else {
        return 0;
    }
}

int isEMPTY() {
    if (st.top == -1) {
        return 1;
    } else {
        return 0;
    }
}

int stack_size() {
    return st.top;
}

int top() {
    return st.items[st.top].m;
}

void push(int m) {
    if (isFULL()) {
        fputs("Error: stack overflow\n", stderr);
        abort();
    } else {
        st.top ++;
        st.items[st.top].m = m;
    }
}

void pop(int *m) {
    if(isEMPTY()) {
        fputs("Error: stack underflow\n", stderr);
        abort();
    } else {
        *m = st.items[st.top].m;
        st.top --;
    }
}

int ackermann(int m, int n) {
    push(m);
    while(!isEMPTY()) {
        pop(&m);
        if(m == 0) {
            n ++;
        } else if(n == 0) {
            n = 1;
            push(m-1);
        } else {
            n --;
            push(m-1);
            push(m);
        }
    }
    return n;
}

int main() {
    int r, m, n;
    struct timeval t1, t2, tv;

    createstack();
    printf("Enter m and n: ");
    scanf("%d%d", &m, &n);

    gettimeofday(&t1,NULL);

    r = ackermann(m,n);

    gettimeofday(&t2,NULL);

    printf("Result = %d\n", r);
    timersub(&t2, &t1, &tv);
    printf("%ld milliseconds\n", (1000000*tv.tv_sec + tv.tv_usec)/1000);

    return 0;
}
